src/main/scala/diplomacy/Nodes.scala

```diff
@@ -5,82 +5,446 @@ package freechips.rocketchip.diplomacy
+  *                  │            │                 │                             │                                        │                     │ │
+  *                  │            │                 │                             │                                        │                     │ │
+  * (immobilize after elaboration)│                 ↓                             │                                        │                     │ │
+  *     [[OutwardNode.oBindings]]─┘      [[MixedNode.oDirectPorts]]───→[[MixedNode.oPorts]]                     [[MixedNode.doParams]]           │ │
+  *                  ↑               (inward port from [[OutwardNode]])           │                                        │                     │ │
+  *                  │                  ┌─────────────────────────────────────────┤                                        │                     │ │
+  *                  │                  │                                         │                                        │                     │ │
+  *                  │                  │                                         │                                        │                     │ │
+  *     [[OutwardNode.accPO]]           │                                         ↓                                        │                     │ │
+  *   (binding node when elaboration)   │ [[MixedNode.diParams]]──────→[[MixedNode.mapParamsD]]────────────────────────────┘                     │ │
+  *                                     │            ↑                                                                                           │ │
+  *                                     │            └───────────────────────────────────────────────────────────────────────────────────────────┘ │
+  *                                     └──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+  *
+  * # Inward/Outward vs. Upward/Downward
+  *
+  * Diplomacy defines two dimensions: inward/outward and upward/downward.
+  *
+  * Inward/outward refer to the direction of the directed graph itself. For a
+  * given node:
+  * - Inward refers to edges that point into a node.
+  * - Outward refers to edges that point out of a node.
+  *
+  * A useful mnemonic for distinguishing between inward and outward is to always
+  * view it from the perspective of a particular node. Because Diplomacy
+  * describes directed, acyclic graphs, this direction will always be consistent
+  * across the entire graph.
+  *
+  * Upward/downward refer to the direction of the parameter negotiation, where
+  * the direction is relative to the inward/outward direction. For a given edge:
+  * - Upward refers to a flow of parameters in the outward direction.
+  * - Downward refers to a flow of parameters in the inward direction.
```

Comment:
The use of inward/outward here is kind of confusing, since we want to distinguish between up/down versus in/out.